### PR TITLE
Use date/time glue patterns for the components bag 

### LINKED ISSUE #585: In skeleton matching, combine date and time patterns
https://unicode.org/reports/tr35/tr35-dates.html#Matching_Skeletons

> 1. Divide the request into a date fields part and a time fields part.
> 2. For each part, find the matching dateFormatItem, and expand the pattern as above.
> 3. Combine the patterns for the two dateFormatItems using the appropriate dateTimeFormat pattern, determined as follows from the requested date fields:
>   * If the requested date fields include wide month (MMMM, LLLL) and weekday name of any length (e.g. E, EEEE, c, cccc), use <dateTimeFormatLength type="full">
>   * Otherwise, if the requested date fields include wide month, use <dateTimeFormatLength type="long">
>   * Otherwise, if the requested date fields include abbreviated month (MMM, LLL), use <dateTimeFormatLength type="medium">
>   * Otherwise use <dateTimeFormatLength type="short">

This is a follow-up to #481. It blocks #272.

### PR DESCRIPTION
**Pull request by gregtatum**
Resolves #585.

Edit: This is now ready for review.

### CODE REVIEW

**Review by zbraniecki — APPROVED** [review]
small stylistic suggestions, optional.

File: components/datetime/src/skeleton.rs [hunk 1]
```diff
@@ -348,12 +349,165 @@ const REQUESTED_SYMBOL_MISSING: u32 = 10000;
...
+        // ... etc.
+        unimplemented!(
+            "There are no \"other\" fields supported, these need to be appended to the pattern."
+        );
+    }
+
+    if date.is_empty() || time.is_empty() {
+        return match first_pattern_match {
+            BestSkeleton::AllFieldsMatch(_) => {
+                panic!("Logic error in implementation. AllFieldsMatch handled above.")
```
[the remark this answers is not in the document]
Comment (gregtatum) [reply components/datetime/src/skeleton.rs:388]: Ah, TIL `unreachable!()`.

I don't think I can do the latter suggestion as the inner pattern changes from a `&Pattern` to `Pattern`.

File: components/datetime/src/skeleton.rs [hunk 2]
```diff
@@ -348,12 +349,165 @@ const REQUESTED_SYMBOL_MISSING: u32 = 10000;
...
+            // TODO(#418)
+            // FieldSymbol::TimeZone(_) => &mut other,
+            // TODO(#486)
+            // FieldSymbol::Era(_) => &mut other,
+            // Plus others...
+        };
+        vec.push(*field)
+    }
+
+    FieldsByType { date, time, other }
```
[the remark this answers is not in the document]
Comment (gregtatum) [reply components/datetime/src/skeleton.rs:512]: I opted something in the middle of the two.

**Review by zbraniecki — APPROVED** [review]

File: components/datetime/src/skeleton.rs [hunk 3]
```diff
@@ -348,12 +349,169 @@ const REQUESTED_SYMBOL_MISSING: u32 = 10000;
...
-pub enum BestSkeleton<'a> {
-    AllFieldsMatch(AvailableFormatPattern<'a>),
-    MissingOrExtraFields(AvailableFormatPattern<'a>),
+pub enum BestSkeleton<T> {
+    AllFieldsMatch(T),
+    MissingOrExtraFields(T),
     NoMatch,
 }
 
+// TODO - This could return a Cow<'a, Pattern>, but it affects every other part of the API to
```
Comment (sffc) [inline components/datetime/src/skeleton.rs:358]: Hopefully Zibi will be able to fix this in his PR to make it such that patterns in DateTimeFormat can always be zero-copy.

**Review by sffc — APPROVED** [review]
All seems plausible

**Pull request merged**